### PR TITLE
feat: Add support for renaming workspaces

### LINKED ISSUE #3000: Allow workspace names to be editable
## Problem 
I want to move my primary workspace off the `coder` template to the `coder-ts` template that Ammar created. The only problem is I obviously can't use my primary workspace name `dev` for both workspaces. 

## Definition of done
It would be nice if I could rename my obsolete workspace and claim `dev` for my new workspace.

### PR DESCRIPTION
**Pull request by mafredri**
**WIP:** This PR adds support for renaming workspaces.

For now, only the `docker` template has been updated, more to follow.

#### TODO

- [x] Prevent/warn about data loss mentioned in #3000, #3386
	- The rename cli command now has a disclaimer
- [ ] ~~Update all templates to use workspace IDs instead of names~~ We will do this separately.

Fixes #3000

### CODE REVIEW

**Comment by Emyrk** [comment]
Do you think we'll need to reuse the pg error checks? Should we have database pkg functions like:

```golang
if database.IsUniqueConstraint(err, "column_name") {
	// ...
}
``` 

Just spitballing ideas. I think you also need this check on the `post` workspace route too right?

**Comment by mafredri** [comment]
@Emyrk I like that suggestion. What do you think about also pulling out the constraints from `dump.sql` and creating constants for them in a generated file?

**Comment by Emyrk** [comment]
> @Emyrk I like that suggestion. What do you think about also pulling out the constraint names from `dump.sql` and creating constants for them in a generated file?
> 
> Edit: Example in [`85eb0cf`](https://github.com/coder/coder/pull/3409/commits/85eb0cfa7fb549e1f1b2d7f80785d0254d10b105).

I've been wondering if we should generate some Golang code like we do typescript code from our database schema. It might be a bit preemptive to go that far now, unless you think this error checking will be useful outside of just workspace and user names.

----

If we do generate though, I'd like to expand our enum generation in Go. Currently sqlc generates Enums to `models.go`. I'd like to expand that to copy the enums to `codersdk` somehow and add some functions for validating enums. Right now you can see we handle some enum validation manually https://github.com/coder/coder/blob/44423b3b28fd57fa3150420b8704e25353c6abc9/codersdk/parameters.go#L98-L98. But this is annoying to maintain in json tags, it would be great to handle this with some autogen "validate" function imo.

**Comment by mafredri** [comment]
@Emyrk 

> I've been wondering if we should generate some Golang code like we do typescript code from our database schema. It might be a bit preemptive to go that far now, **unless you think this error checking will be useful outside of just workspace and user names.**

My thinking is that the value is in ensuring our code is in sync with the schema. For instance, the unique index we're relying on here could be recreated with a different name at some point, and in that case the code would silently no longer catch the case. By generating the enums we ensure that the project fails to build unless names are in sync.

Ultimately it'd be great if we strive for having all relevant parts generated into Go code (to ensure synchronization), but I don't think it needs to be an all or nothing approach. I.e. it'd be fine to generate a subset now, and more as the need arises.

It didn't look like `sqlc` supports generation of unique constraint enums, so I was thinking of doing some manual lifting in this case (based on `dump.sql`).

> But this is annoying to maintain in json tags, it would be great to handle this with some autogen "validate" function imo.

(Not relevant for this PR, but) That looks annoying indeed. I think one of the libraries I mentioned in https://github.com/coder/coder/pull/3390#issuecomment-1209255220 would support the custom `Validate() error` function, which sounds like what we'd want.

**Comment by mafredri** [comment]
While #3386 is not fixed yet, I opted to add a large disclaimer to the rename command:

```
❯ ./coder rename test byebye
> WARNING: A rename can result in loss of home volume if the template references the workspace name. Continue? (yes/no) no
```

I think this should be sufficient for us to merge this functionality.

**Review by presleyp — APPROVED** [review]
Frontend ✅

File: cli/rename.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+			}
+			workspace, err := namedWorkspace(cmd, client, args[0])
+			if err != nil {
+				return xerrors.Errorf("get workspace: %w", err)
+			}
+
+			_, err = cliui.Prompt(cmd, cliui.PromptOptions{
+				Text:      "WARNING: A rename can result in loss of home volume if the template references the workspace name. Continue?",
+				IsConfirm: true,
+			})
```
Comment (johnstcn) [inline cli/rename.go:30]: suggestion: require the user to enter the current name of the workspace to continue

Comment (mafredri) [reply cli/rename.go:30]: That won't be skippable, but yeah, that could be a good safety addition. 👍🏻

File: coderd/database/databasefake/databasefake.go [hunk 2]
```diff
@@ -2090,6 +2091,32 @@ func (q *fakeQuerier) UpdateProvisionerJobWithCompleteByID(_ context.Context, ar
...
+	for i, workspace := range q.workspaces {
+		if workspace.Deleted || workspace.ID != arg.ID {
+			continue
+		}
+		for _, other := range q.workspaces {
+			if other.Deleted || other.ID == workspace.ID || workspace.OwnerID != other.OwnerID {
+				continue
+			}
+			if other.Name == arg.Name {
+				return database.Workspace{}, &pq.Error{Code: "23505", Message: "duplicate key value violates unique constraint"}
```
Comment (johnstcn) [inline coderd/database/databasefake/databasefake.go:2107]: TIL you can just return the correct SQL error codes 👍

Comment (mafredri) [reply coderd/database/databasefake/databasefake.go:2107]: It's a bit janky 😅.. but at least the codes will never change.

File: cli/rename.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	"github.com/coder/coder/cli/cliui"
+	"github.com/coder/coder/codersdk"
+)
+
+func rename() *cobra.Command {
+	cmd := &cobra.Command{
+		Annotations: workspaceCommand,
+		Use:         "rename <workspace> <new name>",
+		Short:       "Rename a workspace",
+		Args:        cobra.ExactArgs(2),
```
Comment (johnstcn) [inline cli/rename.go:16]: We should hide this command until it's 100% safe.

**Review by johnstcn — APPROVED** [review]
I'm OK with this right now, bearing in mind that there is a really big open question about how to do this completely 100% safely. We should hide the command until this is 100% safe to do.

**Comment by mafredri** [comment]
image

**Pull request merged**